hadml/datamodules/gan_datamodule.py

```diff
@@ -251,3 +250,141 @@ def test_dataloader(self):
             ),
         }
         return CombinedLoader(loaders, mode="max_size_cycle")
+
+
+class MultiHadronEventGANDataModule(LightningDataModule):
+    def __init__(
+            self, 
+            data_dir="data/Herwig",
+            raw_file_list=["AllClusters_10K.dat"],
+            processed_filename="herwig_multihadron_events_10K.npy",
+            pid_map_file="pid_to_idx.pkl",
+            train_val_test_split: Tuple[float, float, float] = (0.7, 0.15, 0.15),
+            batch_size=32,
+            num_workers=8,
+            initialise_data_preparation=False,
+            debug=True
+        ):
+        super().__init__()
+        self.processed_filename = os.path.join(os.path.normpath(data_dir),
+                                               "processed", processed_filename)
+        self.train_val_test_split = train_val_test_split
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.data_train: Optional[torch.utils.data.Dataset] = None
+        self.data_val: Optional[torch.utils.data.Dataset] = None
+        self.data_test: Optional[torch.utils.data.Dataset] = None
+
+        parser = HerwigMultiHadronEventParser(
+            data_dir=data_dir,
+            raw_file_list=raw_file_list,
+            processed_filename=processed_filename,
+            pid_map_file=pid_map_file,
+            debug=debug
+        )
+        parser.parse_data()
+
+        print('\n', ' '*20, self.__class__.__name__,'\n', '-'*70, sep='')
+
+        if initialise_data_preparation:
+            self.prepare_data()
+            self.setup()
+            one_train_batch = next(iter(self.train_dataloader()))
+            gen_input, disc_input = one_train_batch
+            print(f"\nGenerator input batch: {len(gen_input)} ([{gen_input.size()}])")
+            print(f"Discriminator input batch: {len(disc_input)} ([{disc_input.size()}])")
+            print("\nGenerator input sample", gen_input[0])
+            print("\nDiscriminator input sample", disc_input[0])
+
+
+    def prepare_data(self):
+        # Load data prepared by the parser
+        with open(self.processed_filename, "rb") as f:
+            data = np.load(f, allow_pickle=True)
+        
+        cluster_kin = data.item()["cluster_kin"]
+        hadron_kin = data.item()["had_kin"]
+        had_type_indices = data.item()["had_type_indices"]
+        cluster_labels = data.item()["cluster_labels"]
+        n_events = len(cluster_kin)
+        self.n_had_types = data.item()["n_had_type_indices"] + 1 # 1 extra type for a stop token
+ 
+        # Assigning hadrons to clusters 
+        n_clusters_extracted_from_events = 0
+        self.clusters, self.hadrons_with_types = [], []
+
+        for i in range(n_events):
+            cluster_in_event = []  # [cluster ... cluster]
+            hadrons_in_event = []  # [ [[hadron ... hadron][h_id ... h_id]] ... [[...][...]] ]
```

Comment:
I'm not saying we have to change something here, but let's discuss whether we want to use dataclasses etc. at one of our meetings